### PR TITLE
Session inflight sub/leave request tracking

### PR DESCRIPTION
**Pull request by aforge**
Should address https://github.com/tinode/chat/issues/535

### CODE REVIEW

File: server/topic_proxy.go [hunk 1]
```diff
@@ -30,34 +30,15 @@ func (t *Topic) runProxy(hub *Hub) {
...
-				log.Println("proxy topic: route broadcast request from proxy to master failed:", err)
-			}
-			if len(t.sessions) == 0 {
-				// No more sessions attached. Start the countdown.
-				killTimer.Reset(idleProxyTopicTimeout)
+		case leave := <-t.unreg:
+			t.handleProxyLeaveRequest(leave, killTimer)
+			if leave.pkt != nil && leave.sess.inflightReqs != nil {
+				// If it's a client initiated request.
+				leave.sess.inflightReqs.Done()
```
Comment (or-else) [inline server/topic_proxy.go:41]: I think it should be `Add(-1)` here. This is a multiplexing session. The leave is for a single user within the session. Also I think `handleProxyLeaveRequest` should return success/fail or the result of `t.remSession(leave.sess, asUid)`.

Comment (aforge) [reply server/topic_proxy.go:43]: `Add(-1)` is equivalent to `Done()`.
Modified `t.handleProxyLeaveRequest()`.

File: server/session.go [hunk 2]
```diff
@@ -520,6 +531,7 @@ func (s *Session) leave(msg *ClientComMessage) {
 		} else {
 			// Unlink from topic, topic will send a reply.
 			s.delSub(msg.RcptTo)
+			s.inflightReqs.Add(1)
```
[the remark this answers is not in the document]
Comment (or-else) [reply server/session.go:534]: Ah, you are counting both subscribes and unsubscribes. Why do you need to count unsubscribes?

File: server/topic.go [hunk 3]
```diff
@@ -306,8 +306,15 @@ func (t *Topic) runLocal(hub *Hub) {
 					log.Printf("topic[%s] subscription failed %v, sid=%s", t.name, err, join.sess.sid)
 				}
 			}
+			if join.sess.inflightReqs != nil {
```
Comment (or-else) [inline server/topic.go:309]: How can it be `nil`?

File: server/topic_proxy.go (hunk 1 above)
Comment (or-else) [reply server/topic_proxy.go:43]: Good point. I was under a wrong impression that `Done()` sets lock count to 0.

I don't see how `leave.sess.inflightReqs` can be zero.

**Review by or-else — APPROVED** [review]

File: server/session.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aforge) [reply server/session.go:534]: Unsubscribes also modify the Topic state - so make sure we don't end up with race conditions.

File: server/topic.go (hunk 3 above)
Comment (aforge) [reply server/topic.go:309]: Multiplexed topic master (short lived) sessions don't have it yet. I should perhaps add support for these too as they modify the master topic state.

**Pull request merged**